File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int l=0; 
        int maxlength =  0;
        unordered_set<char> seen;
        for(int r=0; r<n; r++) {
            while(seen.count(s[r])){
                seen.erase(s[l]);
                l++;
            }
            seen.insert(s[r]);
            int length =  r-l+1;
            maxlength = max(maxlength,length);
        }
        
        return maxlength;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxlen = Integer.MIN_VALUE;
        int l=0, r=0;
        HashSet<Character> hs = new HashSet<>();
        while(r<n) {
           while(hs.contains(s.charAt(r))) {
             hs.remove(s.charAt(l));
             l++;
           }
           hs.add(s.charAt(r));
           maxlen = Math.max(maxlen, r-l+1);
           r++;
        }
        return maxlen == Integer.MIN_VALUE ? 0:maxlen;
    }
}